### ISSUE TITLE
Diagnostics: More General Intervals

### ISSUE THREAD

**Issue by ax3l** [ECP-WarpX/WarpX#744]
We discussed in the last months to make ideally all `_int` intervals in inputs more general.
The idea is to be able to pass individual steps and multiple intervals for varying output.

In the past, we just used a union of intervals with start/stop/stride, e.g. [here](https://picongpu.readthedocs.io/en/latest/usage/plugins.html#period-syntax). Possible for WarpX could be a similar string syntax, e.g.
```
<something>.<diag>_int = 0:10000:500 6000:8000:100 7777 8888
```
Will create diagnostics at `0,500,1'000,1'500...10'000`, as well as `6'000,6'100...8'000` in higher resolution (doubly occurring time steps only dumped once), as well as at steps `7'777` and `8'888`.

Separators (space and colon or comma and semicolon) are of course fully arbitrary and can be set to what our input file parser supports.

As an orthogonal task, we currently unify the interfaces of all diagnostics. Maybe `_int` becomes `_period` or `_interval` for better readability.

cc @MaxThevenet @RevathiJambunathan @lucafedeli88 @RemiLehe @NeilZaim, Henri Vincenti, et al.

**Comment by MaxThevenet** [comment]
I think this is just a fraction of what's needed for diagnostics. It would probably be useful to plot, for instance, only 1% of particles every 100 iterations, and 100% every 1000 iterations etc. In general, let us try and make the diagnostics more modular, so that we can have these types of options.

**Comment by ax3l** [comment]
Yes, I think those are orthogonal tasks. The interval parsing can be implemented in a modular manner already. It might be useful outside of diagnostics, too.

**Comment by ax3l** [comment]
How one can implement the parser: https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/pmacc/pluginSystem/toTimeSlice.hpp

**Comment by MaxThevenet** [comment]
A few points to help for the implementation:
- The AMReX parser that reads input parameters is `ParmParse` (see `amrex/Src/Base/AMReX_ParmParse.H`). Before implementing something in WarpX, let us consider what capabilities we would need, and decide with the AMReX people what should be in WarpX and what belongs to AMReX.
- The new diagnostics are in the `Diagnostics` class, see `Diagnostics/Diagnostics.H` in `warpx/Source/`. Files `Diagnostics/FieldIO.cpp` and `Diagnostics/WarpXIO.cpp` should be deleted soon.
- As mentioned by Axel during the CEA meeting, this is a general capability, not relevant for diagnostics only, so it belongs to `Source/Utils/`. If it's very small it could be written in `WarpXUtil.H/cpp`, otherwise let's create a new file in `Source/Utils/`.

**Comment by NeilZaim** [comment]
Closing this issue since this was implemented in #1000
